### PR TITLE
pallet-ibc integration test 

### PR DESCRIPTION
**Pull request by seunlanlege**
closes ComposableFi/centauri#67
closes ComposableFi/centauri#69
closes ComposableFi/centauri#74 
closes ComposableFi/centauri#71
closes ComposableFi/centauri#79

### CODE REVIEW

File: frame/ibc/ping/src/lib.rs [hunk 1]
```diff
@@ -147,6 +140,21 @@ pub mod pallet {
 	}
 }
 
+impl<T: Config> Pallet<T> {
+	pub fn send_ping_impl(params: SendPingParams) -> Result<(), ibc_primitives::Error> {
```
Comment (blasrodri) [inline frame/ibc/ping/src/lib.rs:144]: where's the suffix `impl` coming from?

File: frame/ibc/primitives/src/runtime_interface.rs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+	use ripemd::Digest;
+	let mut hasher = ripemd::Ripemd160::new();
+	hasher.update(message);
+	let hash = hasher.finalize();
+	let mut res = [0u8; 20];
+	res.copy_from_slice(&hash);
+	res
+}
+
+pub fn ss58_to_account_id_32(raw_str: &str) -> Result<[u8; 32], SS58CodecError> {
```
Comment (blasrodri) [inline frame/ibc/primitives/src/runtime_interface.rs:67]: what's `ss`?

File: frame/ibc/src/lib.rs [hunk 3]
```diff
@@ -345,118 +313,64 @@ pub mod pallet {
...
 			let mut ctx = routing::Context::<T>::new();
-			let result =
-				ibc::core::ics26_routing::handler::deliver::<_, HostFunctions>(&mut ctx, msg)
-					.map_err(|_| Error::<T>::ProcessingError)?;
-			Self::deposit_event(result.events.into());
+			let messages = messages
+				.into_iter()
+				.filter_map(|message| {
+					let type_url = String::from_utf8(message.type_url.clone()).ok()?;
+					let is_permissioned = matches!(
```
Comment (blasrodri) [inline frame/ibc/src/lib.rs:360]: why there's only need to have one permission attribute and not all?

File: frame/ibc/ping/src/lib.rs (hunk 1 above)
Comment (seunlanlege) [reply frame/ibc/ping/src/lib.rs:144]: not sure what you mean?

File: frame/ibc/primitives/src/runtime_interface.rs (hunk 2 above)
Comment (seunlanlege) [reply frame/ibc/primitives/src/runtime_interface.rs:67]: its ss58

File: frame/ibc/src/lib.rs (hunk 3 above)
Comment (seunlanlege) [reply frame/ibc/src/lib.rs:360]: ref: https://github.com/ComposableFi/ibc-rs/issues/31

File: frame/ibc/ping/src/lib.rs (hunk 1 above)
Comment (blasrodri) [reply frame/ibc/ping/src/lib.rs:144]: I was just wondering why the `_impl` on the name of the method - curiosity. Is it something we choose? Or are we following  a convention from ... ??

File: frame/ibc/primitives/src/runtime_interface.rs (hunk 2 above)
Comment (blasrodri) [reply frame/ibc/primitives/src/runtime_interface.rs:67]: ok, just for the ones (like me) that know nothing about this: https://wiki.polkadot.network/docs/build-ss58-registry

File: frame/ibc/src/lib.rs (hunk 3 above)
Comment (blasrodri) [reply frame/ibc/src/lib.rs:360]: I'm not arguing against verifying permissions. My question was regarding the permissions that were being verified. Seems that "as long as it has one" it's ok.

Comment (seunlanlege) [reply frame/ibc/src/lib.rs:360]: It's a filter_map operation. So we filter out permissioned packets.

Comment (blasrodri) [reply frame/ibc/src/lib.rs:360]: got it, thanks

**Review by blasrodri — APPROVED** [review]

File: frame/ibc/src/lib.rs (hunk 3 above)
Comment (blasrodri) [reply frame/ibc/src/lib.rs:360]: I'll just rephrase my question, to be 100% sure:
here we are checking if the message matches _at least_ one permission. That's the definition of having the right to execute the message. However, couldn't it be that a message has rights for something but not for something else? and that we need to check in a more granular way?

Comment (seunlanlege) [reply frame/ibc/src/lib.rs:360]: > here we are checking if the message matches at least one permission

That's actually not what is happening, we collect only packets that require root to be dispatched and dispatch them.

https://github.com/ComposableFi/composable/blob/4fe5537a274ca2186e821a6fd67f4443f6953662/frame/ibc/src/lib.rs#L352-L375

File: frame/ibc/ping/src/lib.rs (hunk 1 above)
Comment (seunlanlege) [reply frame/ibc/ping/src/lib.rs:144]: oh `send_ping` is already used by the pallet call impl

**Review by Wizdave97 — APPROVED** [review]

File: code/parachain/frame/ibc/ping/src/lib.rs [hunk 4]
```diff
@@ -18,15 +18,16 @@ use ibc::{
 	},
 	signer::Signer,
 };
-use ibc_trait::CallbackWeight;
-use scale_info::prelude::string::{String, ToString};
-use sp_std::{marker::PhantomData, prelude::*, vec};
+use ibc_primitives::{port_id_from_bytes, CallbackWeight, IbcHandler, SendPacketData};
+use scale_info::prelude::{format, string::String};
```
Comment (Wizdave97) [inline code/parachain/frame/ibc/ping/src/lib.rs:22]:
```suggestion
use scale_info::prelude::{format, string::{String, ToString}}
```
Minor fix here

File: code/Cargo.toml [hunk 5]
```diff
@@ -34,11 +34,8 @@ members = [
   "parachain/runtime/common",
   "parachain/runtime/primitives",
   "parachain/runtime/picasso",
-  "parachain/runtime/picasso-wasm",
```
[the remark this answers is not in the document]
Comment (seunlanlege) [reply code/Cargo.toml:37]: can't build with cargo

[the remark this answers is not in the document]
Comment (seunlanlege) [reply code/Cargo.toml:37]: We can bring this back, but only if we set up a makefile for this custom build configuration.

**Review by Martin-Kiss — APPROVED** [review]

File: code/parachain/frame/ibc/Cargo.toml [hunk 6]
```diff
@@ -143,6 +143,7 @@ runtime-benchmarks = [
   "tendermint",
   "hex-literal",
   "pallet-timestamp",
+  "pallet-timestamp/runtime-benchmarks",
```
Comment (benluelo) [inline code/parachain/frame/ibc/Cargo.toml:146]: both of these aren't required, `pallet-timestamp/runtime-benchmarks` brings in `pallet-timestamp`

File: .config/dictionaries/technical.dic [hunk 7]
```diff
@@ -56,6 +57,7 @@ assetsinfo
 autofetch
 autorenew
 azuretools
+bbcccccc
```
Comment (KaiserKarel) [inline .config/dictionaries/technical.dic:60]: Same, why are we adding this?

File: .config/dictionaries/technical.dic [hunk 8]
```diff
@@ -307,6 +314,7 @@ tsdx
 tungstenite
 typegen
 typeorm
+uatom
```
Comment (KaiserKarel) [inline .config/dictionaries/technical.dic:317]: What is an uatom?

File: .github/workflows/composable-centauri.yml [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+          password: ${{ secrets.DOCKER_HUB_ACCESS_TOKEN }}
+      - run: |
+          # enable required features (see https://github.com/cachix/install-nix-action/issues/19)
+          mkdir -p ~/.config/nix
+          echo "system-features = kvm" >> ~/.config/nix/nix.conf
+      - run: |
+          cachix watch-exec --jobs 16 --compression-level $CACHIX_COMPRESSION_LEVEL composable-community nix -- build .#bridge-devnet-dali-container --no-update-lock-file --show-trace -L
+          docker load --input ./result
+          docker tag composable-centauri-devnet-container:latest "${{ env.DOCKER_REGISTRY_NAME }}/composable-centauri:latest"
+          docker push "${{ env.DOCKER_REGISTRY_NAME }}/composable-centauri:latest"
```
Comment (KaiserKarel) [inline .github/workflows/composable-centauri.yml:66]: How does this differ from `composable-centauri-publish`?

File: .github/workflows/composable-centauri.yml [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  CACHIX_COMPOSABLE: composable-community
+  CACHIX_COMPRESSION_LEVEL: 3
+
+  DOCKER_REGISTRY_NAME: composablefi
+  # weird, next is implicit dependency of docker action
+  DOCKER_USER_OPTION: '$UID:$GID'
+
+jobs:
+  composable-centauri-publish:
+    # if: ${{ github.ref == 'refs/heads/centauri' }}
```
Comment (KaiserKarel) [inline .github/workflows/composable-centauri.yml:25]: remove deadcode

File: code/parachain/frame/composable-traits/src/vesting.rs [hunk 11]
```diff
@@ -114,12 +113,11 @@ pub struct VestingSchedule<VestingScheduleId, BlockNumber, Moment, Balance: HasC
...
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
-pub struct VestingScheduleInfo<BlockNumber, Moment, Balance: HasCompact> {
+pub struct VestingScheduleInfo<BlockNumber, Moment, Balance> {
 	pub window: VestingWindow<BlockNumber, Moment>,
 	/// Number of vest
 	pub period_count: u32,
 	/// Amount of tokens to release per vest
-	#[codec(compact)]
 	pub per_period: Balance,
 }
```
Comment (KaiserKarel) [inline code/parachain/frame/composable-traits/src/vesting.rs:123]: Seems like unrelated to the IBC PR?

File: .config/dictionaries/technical.dic (hunk 8 above)
Comment (blasrodri) [reply .config/dictionaries/technical.dic:317]: it's part of benchmarks. Not sure if it's intended or a typo @seunlanlege ?
```rs
let denom = "transfer/channel-15/uatom";
```

File: .config/dictionaries/technical.dic (hunk 7 above)
Comment (blasrodri) [reply .config/dictionaries/technical.dic:60]: These are some docs coming from `sp-core` talking about bit manipulations and prefixes in `from_ss58check_with_version`
